File: hrv_calculation_typescript/src/validate_hrv.ts
import { MongoClient } from "mongodb";

async function validateHRV() {
  const client = await MongoClient.connect("mongodb://localhost:27017");
  const db = client.db("phr-db");
  const collection = db.collection("PHR");

  const requiredKeys = ["rmssd", "sdnn", "start_index", "start_timestamp"];
  const docs = await collection.find({ "entries.type": "heartrate" }).toArray();

  let checked = 0;
  let valid = 0;
  let invalid = 0;

  for (const doc of docs) {
    for (const entry of doc.entries) {
      if (entry.type !== "heartrate") continue;
      checked++;

      const hrv = entry.hrv;
      if (!Array.isArray(hrv)) {
        console.error(`[DOC ${doc._id}] No HRV array.`);
        invalid++;
        continue;
      }

      const first = hrv[0];
      if (!first || !requiredKeys.every((k) => k in first)) {
        console.error(`[DOC ${doc._id}] HRV entry missing keys.`);
        invalid++;
        continue;
      }

      valid++;
    }
  }

  console.log("\n HRV Field Validation Summary:");
  console.log(`Documents checked: ${checked}`);
  console.log(`Valid entries    : ${valid}`);
  console.log(`Invalid entries  : ${invalid}`);

  await client.close();
}

validateHRV();

File: hrv_calculation_python/validate_hrv.py
from pymongo import MongoClient


def validate_hrv_fields(mongo_uri="mongodb://localhost:27017", db_name="phr-db", collection_name="PHR"):
    '''
    Validate the HRV fields in PHR documents in MongoDB.
    Args:
        mongo_uri (str): MongoDB connection URI.
        db_name (str): Name of the database.
        collection_name (str): Name of the collection containing PHR documents.
    '''
    # Connect to MongoDB and validate HRV fields
    required_keys = {"rmssd", "sdnn", "start_index", "start_timestamp"}
    client = MongoClient(mongo_uri)
    db = client[db_name]
    phr_collection = db[collection_name]

    docs_checked = 0
    valid_entries = 0
    invalid_entries = 0

    phr_docs = phr_collection.find({"entries.type": "heartrate"})

    for doc in phr_docs:
        entries = doc.get("entries", [])
        for entry in entries:
            if entry.get("type") != "heartrate":
                continue

            hrv = entry.get("hrv")
            docs_checked += 1

            if not isinstance(hrv, list):
                print(f"[DOC {doc['_id']}]'hrv' field is missing or not a list.", flush=True)
                invalid_entries += 1
                continue

            for i, hrv_item in enumerate(hrv):
                if not isinstance(hrv_item, dict):
                    print(f"[DOC {doc['_id']}] hrv[{i}] is not a dictionary.", flush=True)
                    invalid_entries += 1
                    break
                missing = required_keys - hrv_item.keys()
                if missing:
                    print(f"[DOC {doc['_id']}] hrv[{i}] missing fields: {missing}", flush=True)
                    invalid_entries += 1
                    break
            else:
                valid_entries += 1

    print("\n HRV Field Validation Summary:", flush=True)
    print(f"Documents checked: {docs_checked}", flush=True)
    print(f"Valid entries    : {valid_entries}", flush=True)
    print(f"Invalid entries  : {invalid_entries}", flush=True)

if __name__ == "__main__":
    validate_hrv_fields()